### PR TITLE
feat: APP-2978 - Implement Wallet module component

### PR DESCRIPTION
**Pull request by thekidnamedkd**
## Description

Implement Wallet module component with tests & stories

Task: [APP-2798](https://aragonassociation.atlassian.net/browse/APP-2798)

## Type of change

-   [x] New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] I have selected the correct base branch.
-   [x] I have performed a self-review of my own code.
-   [x] I have commented my code, particularly in hard-to-understand areas.
-   [x] I have made corresponding changes to the documentation.
-   [x] My changes generate no new warnings.
-   [x] Any dependent changes have been merged and published in downstream modules.
-   [x] I ran all tests with success and extended them if necessary.
-   [x] I have updated the `CHANGELOG.md` file in the root folder of the package after the [UPCOMING] title and before
        the latest version.

[APP-2798]: https://aragonassociation.atlassian.net/browse/APP-2798?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/modules/components/wallet/wallet.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+export interface IWalletProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+    /**
+     * Callback fired when the user connects their wallet.
+     */
+    onConnect?: () => void;
+    /**
+     * Callback fired when the user disconnects their wallet.
+     */
+    onDisconnect?: () => void;
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:16]: Given that we'll probably have a more complex handling for connecting / disconnecting the user on the app (display custom dialog for connect and display user dialog when user is connected), I wouldn't add here the logic of actually connecting and disconnecting the user.
I'm also not sure if we should couple this component to the connected user state on wagmi, I would probably just let the parent component pass the information about the connected user, we could use the `ICompositeAddress` interface to pass the user address and name.
This means that we can update the component properties to only add the `user: ICompositeAddress` property and trigger the onClick property whenever the user clicks on the wallet component

File: src/modules/components/wallet/wallet.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    connector?: 'injected' | 'walletConnect';
+    /**
+     * If selecting WalletConnect as the connector, you must pass WalletConnect project ID. Required for WalletConnect -- see https://docs.walletconnect.com/
+     */
+    projectId?: string;
+    /**
+     * If `true`, the wallet button will be disabled.
+     */
+    disabled?: boolean;
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:28]: The `disabled` property should already be defined on the extended button props

File: src/modules/components/wallet/wallet.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            {isDisconnected && <div>Connect</div>}
+
+            {connectPending && (
+                <>
+                    <div className="hidden md:mr-3 md:flex">
+                        <StateSkeletonBar className="shrink-0" width="64px" size="lg" />
+                    </div>
+                    <StateSkeletonCircular size="lg" />
+                </>
+            )}
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:87]: I believe that the loading state is supposed to be shown when we try to resolve the member avatar or the ens name, could you double check with Selim?

**Review by cgero-eth — COMMENTED** [review]
Thanks for the implementation @thekidnamedkd!

File: src/modules/components/wallet/wallet.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+export interface IWalletProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+    /**
+     * The connected user details.
+     */
+    user: ICompositeAddress;
+    /**
+     * Whether or not the app is connected to a wallet.
+     */
+    isConnected: boolean;
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:16]: If the user is not connected, we don't have an address to set. We can just remove the `isConnected` property, set the `user` property as optional and display the connect button / user avatar-ens depending if the `user` property is set

File: src/modules/components/wallet/wallet.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+export const Wallet: React.FC<IWalletProps> = (props) => {
+    const { user, isConnected = false, onClick, ...otherProps } = props;
+    const { data: ensName, isLoading: ensLoading } = useEnsName({
+        address: isConnected ? addressUtils.getChecksum(user.address) : undefined,
+        query: { enabled: isConnected },
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:23]: The query should only be enabled if the `user.name` property is not already set and if the `user` property is set

File: src/modules/components/wallet/wallet.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    });
+
+    const buttonClassName = classNames(
+        'flex h-12 items-center rounded-full border border-neutral-100 bg-neutral-0 px-4 py-1 text-neutral-500 transition-all',
+        'hover:border-neutral-200',
+        'focus:outline-none focus-visible:ring focus-visible:ring-primary focus-visible:ring-offset',
+        'active:bg-neutral-50 active:text-neutral-800',
+        {
+            '!p-1 md:!pl-4': isConnected,
+        },
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:33]: Let's avoid using the `!important` rule, it's usually a bad practice

File: src/modules/components/wallet/wallet.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+export const Wallet: React.FC<IWalletProps> = (props) => {
+    const { user, isConnected = false, onClick, ...otherProps } = props;
+    const { data: ensName, isLoading: ensLoading } = useEnsName({
+        address: isConnected ? addressUtils.getChecksum(user.address) : undefined,
+        query: { enabled: isConnected },
+    });
+
+    const buttonClassName = classNames(
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:26]: We also need to append an eventual `className` property

File: src/modules/components/wallet/wallet.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        'hover:border-neutral-200',
+        'focus:outline-none focus-visible:ring focus-visible:ring-primary focus-visible:ring-offset',
+        'active:bg-neutral-50 active:text-neutral-800',
+        {
+            '!p-1 md:!pl-4': isConnected,
+        },
+    );
+
+    return (
+        <button className={buttonClassName} onClick={onClick} {...otherProps}>
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:37]: No need to destructure the `onClick` property and pass it, just keep it on the `otherProps` object
```suggestion
        <button className={buttonClassName} {...otherProps}>
```

File: src/modules/components/wallet/wallet.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+            {!isConnected && <div>Connect</div>}
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:38]: The `<div/>` tag is probably redundant
```suggestion
            {!isConnected && 'Connect'}
```

File: src/modules/components/wallet/wallet.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                    <div
+                        title={user.name ?? user.address}
+                        className="hidden min-w-0 max-w-24 items-center truncate md:mr-3 md:flex"
+                    >
+                        {user.name ??
+                            (ensLoading ? (
+                                <StateSkeletonBar size="lg" />
+                            ) : (
+                                ensName ?? addressUtils.truncateAddress(user.address)
+                            ))}
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:51]: `ensLoading` should only be set when actually loading the ENS, we can cleanup the checks here
```suggestion
                            ensLoading ? (
                                <StateSkeletonBar size="lg" />
                            ) : (
                                ensName ?? addressUtils.truncateAddress(user.address)
                            )
```

File: src/modules/components/wallet/wallet.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+ * Example usage of a connected Wallet component with only address.
+ */
+export const WithAddress: Story = {
+    args: {
+        user: {
+            address: '0x1234567890123456789012345678901234567890',
+        },
+        isConnected: true,
+    },
+    render: (props) => <Wallet {...props} />,
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.stories.tsx:38]: Man, these redundant `render` function on the stories 👀

File: src/modules/components/wallet/wallet.test.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        expect(button).toHaveTextContent('Connect');
+    });
+
+    it('renders member avatar when connected', () => {
+        const isConnected = true;
+        createTestComponent({ isConnected });
+        expect(screen.getByTestId('member-avatar-mock')).toBeInTheDocument();
+    });
+
+    it('renders user address when connected', () => {
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.test.tsx:57]:
```suggestion
    it('renders user address when connected and user has no ENS name linked', () => {
```

File: src/modules/components/wallet/wallet.test.tsx [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    it('renders user name when provided and connected', () => {
+        const isConnected = true;
+        const user = {
+            address: '0x0987654321098765432109876543210987654321',
+            name: 'aragon.eth',
+        };
+        createTestComponent({ user, isConnected });
+        expect(screen.getByText('aragon.eth')).toBeInTheDocument();
+        expect(screen.queryByText('0x09…4321')).not.toBeInTheDocument();
+    });
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.test.tsx:80]: Would be nice to have these tests as well:
- renders user ENS when user address has a ENS name linked
- does not fetch user ENS when already provided on the user property

File: src/modules/components/wallet/wallet.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+export const Wallet: React.FC<IWalletProps> = (props) => {
+    const { user, isConnected = false, onClick, ...otherProps } = props;
+    const { data: ensName, isLoading: ensLoading } = useEnsName({
+        address: isConnected ? addressUtils.getChecksum(user.address) : undefined,
+        query: { enabled: isConnected },
+    });
+
+    const buttonClassName = classNames(
+        'flex h-12 items-center rounded-full border border-neutral-100 bg-neutral-0 px-4 py-1 text-neutral-500 transition-all',
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.tsx:27]: As we're already setting a vertical padding, the `h-12` shouldn't be needed

File: src/modules/components/wallet/wallet.test.tsx [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+    it('renders user address when connected and user has no ENS name linked', () => {
+        const user = {
+            address: '0x0987654321098765432109876543210987654321',
+        };
+        jest.spyOn(wagmi, 'useEnsName').mockReturnValue({
+            data: null,
+            isLoading: false,
+        } as wagmi.UseEnsNameReturnType);
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.test.tsx:62]:
```suggestion
        useEnsNameMock.mockReturnValue({
            data: null,
            isLoading: false,
        } as wagmi.UseEnsNameReturnType);
```

File: src/modules/components/wallet/wallet.test.tsx [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+        createTestComponent({ user });
+        expect(screen.getByText('0x09…4321')).toBeInTheDocument();
+    });
+
+    it('renders linked ENS name when connected and no user name provided', () => {
+        const user = {
+            address: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',
+        };
+        jest.spyOn(wagmi, 'useEnsName').mockReturnValue({
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.test.tsx:72]:
```suggestion
        useEnsNameMock.mockReturnValue({
```

File: src/modules/components/wallet/wallet.test.tsx [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        expect(screen.getByText('vitalik.eth')).toBeInTheDocument();
+        expect(screen.queryByText('0xd8…6045')).not.toBeInTheDocument();
+    });
+
+    it('renders user name provided when connected and does not fetch linked ENS name', () => {
+        const user = {
+            address: '0x0987654321098765432109876543210987654321',
+            name: 'vitalik.eth',
+        };
+        jest.spyOn(wagmi, 'useEnsName').mockReturnValue({
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.test.tsx:87]:
```suggestion
        useEnsNameMock.mockReturnValue({
```

**Review by cgero-eth — APPROVED** [review]
Still some minor cleanups on the wallet tests to be made but PR looks good, thanks @thekidnamedkd!

File: src/modules/components/wallet/wallet.tsx (hunk 10 above)
Comment (cgero-eth) [reply src/modules/components/wallet/wallet.tsx:49]: @thekidnamedkd It looks like this suggestion hasn't been applied?

Comment (thekidnamedkd) [reply src/modules/components/wallet/wallet.tsx:49]: But don't we want the order to be "if available" user.name => ensName fetch => truncated address. This suggestion would remove provided user.name as first choice?

Comment (thekidnamedkd) [reply src/modules/components/wallet/wallet.tsx:49]: Also for this reason I thought I understood your suggestion and simplified the outside check.

Comment (cgero-eth) [reply src/modules/components/wallet/wallet.tsx:49]: You're right, the suggestion above does not consider the `user.name` attribute :/ In that case we could cleanup this a bit by doing something like:
```
const truncatedAddress
const displayedName = user?.name ?? ensName ?? addressUtils.truncateAddress(user?.address);

return (
  [...]
  {ensLoading ? <StateSkeletonBar size="lg" /> : displayedName}
```
Would this work?

Comment (thekidnamedkd) [reply src/modules/components/wallet/wallet.tsx:49]: Perfect. This follows some similar patterns for `const resolvedXYZ = ...` elsewhere in ODS then :)

Comment (cgero-eth) [reply src/modules/components/wallet/wallet.tsx:49]: 🙌🏻

File: src/modules/components/wallet/wallet.stories.tsx [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+export const Default: Story = {};
+
+/**
+ * Example usage of a connected Wallet component with only address.
+ */
+export const WithAddress: Story = {
+    args: {
+        user: {
+            address: '0x1234567890123456789012345678901234567890',
+            name: 'TESTTTT',
```
Comment (cgero-eth) [inline src/modules/components/wallet/wallet.stories.tsx:35]:
```suggestion
```

**Review by cgero-eth — APPROVED** [review]

**Pull request merged**